### PR TITLE
[BugFix] Fix ORC late materialization bug when contains complex type column

### PR DESCRIPTION
**Pull request by Smith-Cruise**
ORC ASAN failed / StringVectorBatch oom / wrong answer about nullability may occur when a column contains complex type, and complex type column is treated as lazy load column. 

But our implementation still has some drawbacks.
* We can't support struct subfield lazy load, only support for root node now. (But for now, we don't support lazy materialization for subfields)
* Must do `filterOnFields(data, size, true_size, active_load_orc_positions, false)` before `filterOnFields(data, size, true_size, lazy_load_orc_positions, true)`, otherwise we will set wrong `notNulls` vector. This means that the wrong order of calls will cause bugs. (You must make sure to filter the active column first, then do filter the lazy column)

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc [hunk 1]
```diff
@@ -1022,6 +1022,8 @@ class StructColumnReader : public ColumnReader {
     void nextInternal(const std::vector<std::unique_ptr<ColumnReader>>& children,
                       const std::vector<uint64_t>& fieldIndex, ColumnVectorBatch& rowBatch, uint64_t numValues,
                       char* notNull);
+
+    bool isAllFieldLazy();
```
Comment (dirtysalt) [inline be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc:1026]: AllFields

File: be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc [hunk 2]
```diff
@@ -1245,6 +1295,13 @@ void ListColumnReader::seekToRowGroup(PositionProviderMap* positions) {
     }
 }
 
+void ListColumnReader::lazyLoadSeekToRowGroup(PositionProviderMap* positions) {
+    // If List support lazy load, here need to check
```
Comment (dirtysalt) [inline be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc:1299]: here to check what?

**Comment by dirtysalt** [comment]
> Must do filterOnFields(data, size, true_size, active_load_orc_positions, false) before filterOnFields(data, size, true_size, lazy_load_orc_positions, true), otherwise we will set wrong notNulls vector. This means that the wrong order of calls will cause bugs. (You must make sure to filter the active column first, then do filter the lazy column)

Could we add a flag to check if `(true)` is called before `(false)`?  If not, we can throw exception to find problem ASAP.

File: be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc (hunk 2 above)
Comment (Smith-Cruise) [reply be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc:1299]: typo, just means If we support list/map/struct's lazy load, we need to handle `null` carefully.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc (hunk 1 above)
Comment (Smith-Cruise) [reply be/src/formats/orc/apache-orc/c++/src/ColumnReader.cc]: fixed

**Comment by Smith-Cruise** [comment]
> > Must do filterOnFields(data, size, true_size, active_load_orc_positions, false) before filterOnFields(data, size, true_size, lazy_load_orc_positions, true), otherwise we will set wrong notNulls vector. This means that the wrong order of calls will cause bugs. (You must make sure to filter the active column first, then do filter the lazy column)
> 
> Could we add a flag to check if `(true)` is called before `(false)`? If not, we can throw exception to find problem ASAP.

added

File: be/src/formats/orc/apache-orc/c++/src/Vector.cc [hunk 3]
```diff
@@ -300,8 +300,12 @@ void StructVectorBatch::filter(uint8_t* f_data, uint32_t f_size, uint32_t true_s
 void StructVectorBatch::filterOnFields(uint8_t* f_data, uint32_t f_size, uint32_t true_size,
                                        const std::vector<int>& positions, bool onLazyLoad) {
     if (!onLazyLoad) {
+        alreadyFilteredByActiveColumn = true;
         ColumnVectorBatch::filter(f_data, f_size, true_size);
     } else {
+        if (!alreadyFilteredByActiveColumn) {
+            throw orc::ParseError("We have to filter active column first");
```
Comment (dirtysalt) [inline be/src/formats/orc/apache-orc/c++/src/Vector.cc:307]: "orc parser filters active columns before lazy columns"

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/formats/orc/apache-orc/c++/src/Vector.cc (hunk 3 above)
Comment (Smith-Cruise) [reply be/src/formats/orc/apache-orc/c++/src/Vector.cc]: fixed

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by letian-jiang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by dirtysalt — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
@Mergify backport branch-3.1 branch-3.0 branch-2.5

**Pull request merged**